app/GioForegroundService.java

```diff
@@ -0,0 +1,82 @@
+// SPDX-License-Identifier: Unlicense OR MIT
+
+package org.gioui;
+import android.app.Notification;
+import android.app.Service;
+import android.app.Notification;
+import android.app.Notification.Builder;
+import android.app.NotificationChannel;
+import android.app.NotificationManager;
+import android.app.PendingIntent;
+import android.content.Context;
+import android.content.Intent;
+import android.os.IBinder;
+
+// GioForegroundService implements a Service required to use the FOREGROUND_SERVICE
+// permission on Android, in order to run an application in the background.
+// See https://developer.android.com/guide/components/foreground-services for
+// more details. To add this permission to your application, import
+// gioui.org/app/permission/foreground and use the Start method from that
+// package to control this service.
+public class GioForegroundService extends Service {
+	private String channelID;
+	private String channelName;
+	private String channelDesc;
+	private int notificationID = 0x42424242;
+
+	@Override public int onStartCommand (Intent intent, int flags, int startId) {
+		// get the channel parameters from intent extras
+		channelID   = intent.getStringExtra("channelID");
+		channelName = intent.getStringExtra("channelName");
+		channelDesc = intent.getStringExtra("channelDesc");
+		notificationID = intent.getIntExtra("notificationID", notificationID);
+		this.createNotificationChannel();
+
+		// create the Intent that will bring GioActivity to foreground
+		Context ctx = getApplicationContext();
+		Intent resultIntent = new Intent(ctx, GioActivity.class);
```

Comment:
GioActivity may not be the actual class used when Gio is embedded ala Tailscale. I suggest bundling resultIntent along with channelID, channelName, channelDesc.

Reply:
```
diff --git a/app/Gio.java b/app/Gio.java
index d6db219c..30a13f30 100644
--- a/app/Gio.java
+++ b/app/Gio.java
@@ -2,10 +2,13 @@
 
 package org.gioui;
 
+import android.app.Activity;
 import android.content.ClipboardManager;
 import android.content.ClipData;
 import android.content.Context;
 import android.content.Intent;
+import android.content.pm.PackageManager;
+import android.os.Bundle;
 import android.os.Handler;
 import android.os.Looper;
 
@@ -67,16 +70,25 @@ public final class Gio {
 
        static private native void scheduleMainFuncs();
 
-       static Intent startForegroundService(Context ctx, String serviceClassName, String activityClassName, String title, String text) {
+       static Intent startForegroundService(Context ctx, String title, String text) {
                Intent intent = new Intent();
                try {
-                       intent.setClass(ctx, ctx.getClassLoader().loadClass(serviceClassName));
+                       Activity activity = (Activity) ctx;
+                       Bundle metadata = ctx.getPackageManager().getActivityInfo(activity.getComponentName(), PackageManager.GET_META_DATA).metaData;
+                       if (metadata != null) {
+                               String serviceClassName = metadata.getString("org.gioui.ForegroundServiceClass");
+                               intent.setClass(ctx, ctx.getClassLoader().loadClass(serviceClassName));
+                       } else {
+                               throw new RuntimeException("No ForegroundService MetaData found");
+                       }
                } catch (ClassNotFoundException e) {
                        throw new RuntimeException(e);
+               } catch (PackageManager.NameNotFoundException e) {
+                       throw new RuntimeException(e);
                }
+
                intent.putExtra("title", title);
                intent.putExtra("text", text);
-               intent.putExtra("activityClass", activityClassName);
                ctx.startService(intent);
                return intent;
        }
diff --git a/app/GioForegroundService.java b/app/GioForegroundService.java
index 1bef2e2c..4a587baa 100644
--- a/app/GioForegroundService.java
+++ b/app/GioForegroundService.java
@@ -37,9 +37,10 @@ public class GioForegroundService extends Service {
                                channelName = metadata.getString("org.gioui.ForegroundChannelName");
                                String channelDesc = metadata.getString("org.gioui.ForegroundChannelDesc");
                                String channelID = metadata.getString("org.gioui.ForegroundChannelID");
+                               String resultActivity = metadata.getString("org.gioui.ForegroundResultActvity");
                                int notificationID = metadata.getInt("org.gioui.ForegroundNotificationID", 0x42424242);
                                this.createNotificationChannel(channelDesc, channelID, channelName);
-                               Intent resultIntent = new Intent(ctx, ctx.getClassLoader().loadClass(extras.getString("activityClass")));
+                               Intent resultIntent = new Intent(ctx, ctx.getClassLoader().loadClass(resultActivity));
                                PendingIntent pending = PendingIntent.getActivity(ctx, notificationID, resultIntent, Intent.FLAG_ACTIVITY_CLEAR_TASK);
                                Notification.Builder builder = new Notification.Builder(ctx, channelID)
                                        .setContentTitle(title)
diff --git a/app/os_android.go b/app/os_android.go
index 006b763c..af27e60c 100644
--- a/app/os_android.go
+++ b/app/os_android.go
@@ -969,22 +969,17 @@ var foregroundService struct {
 
 // StartForeground starts the foreground service
 func StartForeground(title, text string) (err error) {
-       const serviceClass = "org/gioui/GioForegroundService"
-       const activityClass = "org/gioui/GioActivity"
-
        foregroundService.mu.Lock()
        defer foregroundService.mu.Unlock()
        foregroundService.once.Do(func() {
                runInJVM(javaVM(), func(env *C.JNIEnv) {
                        startForegroundService := getStaticMethodID(env, android.gioCls,
                                "startForegroundService",
-                               "(Landroid/content/Context;Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;)Landroid/content/Intent;",
+                               "(Landroid/content/Context;Ljava/lang/String;Ljava/lang/String;)Landroid/content/Intent;",
                        )
                        foregroundService.intent, err = callStaticObjectMethod(env, android.gioCls,
                                startForegroundService,
                                jvalue(android.appCtx),
-                               jvalue(javaString(env, serviceClass)),
-                               jvalue(javaString(env, activityClass)),
                                jvalue(javaString(env, title)),
                                jvalue(javaString(env, text)),
                        )
diff --git a/cmd/gogio/androidbuild.go b/cmd/gogio/androidbuild.go
index bb216191..559bb9e3 100644
--- a/cmd/gogio/androidbuild.go
+++ b/cmd/gogio/androidbuild.go
@@ -469,6 +469,9 @@ func exeAndroid(tmpDir string, tools *androidTools, bi *buildInfo, extraJars, pe
                                <action android:name="android.intent.action.MAIN" />
                                <category android:name="android.intent.category.LAUNCHER" />
                        </intent-filter>
+                       <meta-data android:name="org.gioui.ForegroundServiceClass"
+                       android:value="org/gioui/GioForegroundService" />
+
                </activity>
                {{if .HasService}}
                <service android:name="org.gioui.GioForegroundService"
@@ -481,6 +484,8 @@ func exeAndroid(tmpDir string, tools *androidTools, bi *buildInfo, extraJars, pe
                        android:value="ForegroundChannelDesc" />
                        <meta-data android:name="org.gioui.ForegroundNotificationID"
                        android:value="0x42424242" />
+                       <meta-data android:name="org.gioui.ForegroundResultActivity"
+                       android:value="org/gioui/GioActivity" />
                </service>
                {{end}}
        </application>
```
The above diff builds but doesn't work and as far as I can see doesn't drop any error in logcat either. So, I'm a bit confused how to debug this.

Reply:
Above diff applied:
https://github.com/mixmasala/gio/pull/new/wip_foreground_class_metadata

Reply:
What do you mean by doesn't work? Some notes:

- Any exceptions thrown from Java is reflected in the returned error value. Make sure you check them.
- You don't need a meta-data element for ForegroundResultActivity. You can [query the current app's launcher activity](https://stackoverflow.com/questions/14129274/get-instance-of-current-android-intent-category-launcher-activity).
- Why not hard-code ForegroundServiceClass name? I don't see the need to have a meta-data parameter for it.

Reply:
ah nice. thanks!